### ISSUE TITLE
Trouble understanding indexing in loss function, or bug

### ISSUE THREAD

**Issue by gsimchoni** [pierluigiferrari/ssd_keras#21]
(First of course, this is an amazing piece of code, thank you)

In the `compute_loss` function:
```

# 1: Compute the losses for class and box predictions for every box

        classification_loss = tf.to_float(self.log_loss(y_true[:,:,:-12], y_pred[:,:,:-12])) # Output shape: (batch_size, n_boxes)
        localization_loss = tf.to_float(self.smooth_L1_loss(y_true[:,:,-12:-8], y_pred[:,:,-12:-8])) # Output shape: (batch_size, n_boxes)
```

I'm probably wrong, but if there are say 6 classes, then the last dimension of `y_true` should be 6 one-hot encoded classes + 4 ground truth box coordinates + 4 arbitrary entries = 14. And so by indexing `y_true[:,:,-12]` we enter the `log_loss` function only the first 2 one-hot encoded classes, no?

If there were only 2 classes this would throw an error.

Shouldn't it be in fact `y_true[:, :, -8]`?

Same with `localization_loss` which should take the 4 box coordinates, shouldn't it be `y_true[:,:,-8:-4]`?

(Sorry if I'm completely wrong)

**Comment by pierluigiferrari** [comment]
Fortunately you're wrong, otherwise this implementation wouldn't work at all :slightly_smiling_face: 

What was incorrect, however, was the documentation in some classes and functions that claimed that the shape of the model's output tensor is `(batch_size, #boxes, #classes + 8)`, when it is in fact `(batch_size, #boxes, #classes + 12)` (the last four entries are the four variance values), so thank you for raising the issue! I've just corrected the documentation in the places that were affected. The variance values are only relevant for encoding the ground truth and decoding the predictions, however, so the loss function doesn't care about them. So as far as the loss function is concerned, the last 8 entries could just be dummy values.

**Comment by gsimchoni** [comment]
Fortunately! Thank you!
